### DISCUSSION TITLE
logback mdc baggage with root span

### DISCUSSION THREAD

**Discussion by AntonioLyubchev** [open-telemetry/opentelemetry-java-instrumentation#12548] [Q&A]
I am trying to add this Baggage to the context that is part of the local root span. My idea is that at some point in the service, I will have an ID that I want to place in the root span as baggage and I want to propagate it until the end of the trace. Is this achievable? I tried with something like:

``` java
Baggage baggage = Baggage.builder()
        .put("TEST", "abc123")
        .build();

final var rootSpan = LocalRootSpan.current();
rootSpan.storeInContext(Context.current().with(baggage));
```
but the next log messages do not have the baggage values as MDC.

It does work when I do it like this:
``` java
Baggage baggage = Baggage.builder()
        .put("PSO_ID", "abc123")
        .build();

try (var scope = Context.current().with(baggage).makeCurrent()) {

    log.info("test {}", MDC.get("baggage.PSO_ID"));
}
```

but only for one log message, the one in the try:
![image](https://github.com/user-attachments/assets/6ac4850c-c2b3-486c-b4f5-1ada9353cbd5)

I am using 
``` XML
<dependency>
    <groupId>io.opentelemetry.instrumentation</groupId>
    <artifactId>opentelemetry-logback-mdc-1.0</artifactId>
    <version>2.9.0-alpha</version>
</dependency>
```
with spring boot opentelemetry starter

Also, I was expecting to be able to access the MDC values with MDC.get but the context map is empty.

**Comment by AntonioLyubchev** [comment]
An alternative solution to my issue would be a way to add specific Span attributes to every log message further down the trace. So I can set a span attribute at some point and from that point on it will be included in every log message like an MDC, is this doable? In the end I am just looking for a way to set a field once and that field must be present in every log message.

**Comment by laurit** [comment]
For the added baggage to become visible you have to add it to context and then make that context current. In `rootSpan.storeInContext(Context.current().with(baggage));` you are discarding the context returned from `storeInContext` so the current context will remain as it was. Note that what you are doing is weird, local root span is already present in the context, what are you trying to achieve with this?

> Also, I was expecting to be able to access the MDC values with MDC.get but the context map is empty.

The goal of that instrumentation is to expose trace_id, span_id, baggage values etc. so that they could be used in logging patterns. Exposing them in `MDC.get` is not a goal.

> I am trying to add this Baggage to the context that is part of the local root span. My idea is that at some point in the service, I will have an ID that I want to place in the root span as baggage and I want to propagate it until the end of the trace. Is this achievable? I tried with something like:

Baggage is not tied to the span, it is a separate concept. When context propagation to an external service happens baggage propagator looks for the current baggage from the context and propagates it.

**Reply by AntonioLyubchev** [reply]
Thanks for the reply, yes it did feel weird but is there a way for me to achieve what I am trying to do?

I don't want to wrap my kafka producer for example with a scope so the baggage can be propagated. Is there a way that certain span attributes can ALWAYS be injected into the logs?

For example, an the service entry point I can set an IP address in a span attribute, and from this point up until the fifth service in chain -> I want to have that IP address as part of every structured log message.

Having everything in the baggage at all times with combination of the addBaggage seems to solve the issue but I can only use baggage in scopes and I don't want to wrap my whole initial service in a scope

**Reply by laurit** [reply]
The intended way to use otel context is to open a scope by calling `makeCurrent()`, this will make `Context.current()` return that context, and close the scope when you are done with the context. Scopes must be closed, and they must be closed in the order they were opened. Failure to close the scope will breaks stuff in a horrible way.

> For example, an the service entry point I can set an IP address in a span attribute, and from this point up until the fifth service in chain -> I want to have that IP address as part of every structured log message.

You could build a custom baggage propagator reads some attribute from the root span and includes it in the propagated baggage.

**Reply by AntonioLyubchev** [reply]
> You could build a custom baggage propagator reads some attribute from the root span and includes it in the propagated baggage.

This sounds like what  I actually need, can you please point me to a resource or example I would be infinitely grateful! The question is, can I build it server-side or do I need changes in the otel project?

**Reply by AntonioLyubchev** [reply]
Or actually, can I make something that for every log message it will add values from the Otel Context to the structured json log as attributes? That way I can just fill up the Context and log messages will have the state I need

**Reply by trask** [reply]
I think this is what you want (if it existed): https://github.com/open-telemetry/oteps/pull/207
